File: src/main/java/net/orandja/mcutils/NBTUtils.java
package net.orandja.mcutils;

import net.minecraft.nbt.NbtCompound;
import net.minecraft.nbt.NbtElement;
import net.minecraft.nbt.NbtList;
import net.minecraft.nbt.NbtString;

import java.util.List;
import java.util.function.Consumer;

public abstract class NBTUtils {

    public static NbtList toNbtList(List<NbtElement> list) {
        NbtList nbtList = new NbtList();
        nbtList.addAll(list);
        return nbtList;
    }

    public static NbtList addTo(NbtList list, NbtElement element) {
        list.add(element);
        return list;
    }

    public static NbtList addTo(NbtList list, String string) {
        list.add(NbtString.of(string));
        return list;
    }

    public interface NbtSupplier<N> {
        N get(String id);
    }

    public interface NbtCompute<N> {
        N get();
    }

    public static <N extends NbtElement> N getOrCompute(NbtCompound tag, String id, NbtSupplier<N> supplier, NbtCompute<N> compute, Consumer<N> consumer) {
        N value;
        if(tag.contains(id)) {
            value = supplier.get(id);
        } else {
            value = compute.get();
            tag.put(id, value);
        }

        consumer.accept(value);

        return value;
    }

    public static <N extends NbtElement> N getOrCreate(NbtCompound tag, String id, NbtSupplier<N> supplier, NbtCompute<N> compute) {
        if(tag.contains(id)) {
            return (N) tag.get(id);
        } else {
            N value = compute.get();
            tag.put(id, value);
            return value;
        }
    }

    public static NbtCompound getTagOrCompute(NbtCompound tag, String id, Consumer<NbtCompound> consumer) {
        return getOrCompute(tag, id, tag::getCompound, NbtCompound::new, consumer);
    }

    public static NbtList getStringListOrCompute(NbtCompound tag, String id, Consumer<NbtList> consumer) {
        return getStringListOrCompute(tag, id, NbtList::new, consumer);
    }

    public static NbtList getStringListOrCompute(NbtCompound tag, String id, NbtCompute<NbtList> compute, Consumer<NbtList> consumer) {
        return getOrCompute(tag, id, key -> tag.getList(key, NbtElement.STRING_TYPE), compute, consumer);
    }
}

File: src/main/kotlin/net/orandja/mcutils/NBTUtils.kt
package net.orandja.mcutils

import net.minecraft.item.ItemStack
import net.minecraft.nbt.NbtCompound
import net.minecraft.nbt.NbtElement
import net.minecraft.nbt.NbtList
import net.minecraft.nbt.NbtString

fun List<NbtElement>.toNBTList(): NbtList =
    NbtList().apply list@{ this@toNBTList.forEach { this@list.add(it) } }

operator fun NbtList.plus(string: String): NbtList = run {
    this.add(NbtString.of(string))
    return this
}
operator fun NbtList.plus(element: NbtElement): NbtList = run {
    this.add(element)
    return this
}

fun <N: NbtElement> NbtCompound.getOrCompute(id: String, getter: (String) -> N?, compute: () -> N?, consumer: (N) -> Unit): N {
    var value: N?
    if(this.contains(id))
        value = getter(id)
    else {
        value = compute()
        this.put(id, value)
    }

    consumer.invoke(value!!)
    return value
}

fun <N: NbtElement> NbtCompound.getOrCreate(id: String, getter: (String) -> N?, compute: () -> N?): N {
    return if(this.contains(id))
        getter(id)!!
    else
        compute().apply { this@getOrCreate.put(id, this) }!!
}

fun NbtCompound.getTagOrCompute(id: String, consumer: (NbtCompound) -> Unit): NbtCompound {
    return getOrCompute(id, this::getCompound, ::NbtCompound, consumer)
}

fun NbtCompound.getSLOrCompute(id: String, compute: () -> NbtList = ::NbtList, consumer: (NbtList) -> Unit): NbtList {
    return getOrCompute(id, this::getStringList, compute, consumer)
}

fun NbtCompound.getStringList(id: String): NbtList {
    return this.getList(id, 8)
}

fun ItemStack.computeLore(consumer: (NbtList) -> Unit) {
    this.computeTag {
        it.getTagOrCompute("display") { display ->
            display.getSLOrCompute("Lore", ::NbtList, consumer)
        }
    }
}

fun NbtList._addAll(collection: Iterable<NbtString>) = apply { collection.forEach(this::add) }